File: java/perfect-numbers/src/main/java/NaturalNumber.java
class NaturalNumber {
    private final int number;
    private final Classification classification;

    NaturalNumber(int number) {
        this.number = number;

        if (number <= 0) {
            throw new IllegalArgumentException("You must supply a natural number (positive integer)");
        }

        int divisorSum = 1;

        for (int i = 2; i * i <= number; i++) {
            if (number % i == 0) {
                divisorSum += i;

                if (i != number / i) {
                    divisorSum += number / i;
                }
            }
        }

        if (number == 1) {
            divisorSum = 0;
        }

        if (divisorSum < number) {
            this.classification = Classification.DEFICIENT;
        } else if (divisorSum > number) {
            this.classification = Classification.ABUNDANT;
        } else {
            this.classification = Classification.PERFECT;
        }
    }

    Classification getClassification() {
        return classification;
    }
}

File: kotlin/perfect-numbers/src/main/kotlin/NaturalNumber.kt
enum class Classification {
    DEFICIENT, PERFECT, ABUNDANT
}

fun classify(naturalNumber: Int): Classification {
    require(naturalNumber > 0)

    val sum = ((1..(naturalNumber / 2)).filter { i -> naturalNumber % i == 0 }.sum())
    return when {
        sum < naturalNumber -> Classification.DEFICIENT
        sum == naturalNumber -> Classification.PERFECT
        else -> Classification.ABUNDANT
    }
}
